src/components/BoardBar/ExportWhiteboardDialogDownloadFile.tsx

```diff
@@ -14,54 +14,56 @@
  * limitations under the License.
  */
 
+import { useWidgetApi } from '@matrix-widget-toolkit/react';
 import FileDownloadIcon from '@mui/icons-material/FileDownload';
 import { LoadingButton } from '@mui/lab';
-import { PropsWithChildren, useEffect, useState } from 'react';
+import { PropsWithChildren, useCallback, useState } from 'react';
+import { downloadData } from '../../lib';
 import { useActiveWhiteboardInstance } from '../../state';
 import { useGetRoomNameQuery } from '../../store';
 
 export function ExportWhiteboardDialogDownloadFile({
   children,
   onClick,
 }: PropsWithChildren<{ onClick: () => void }>) {
-  const { data: roomNameStateEvent } = useGetRoomNameQuery();
-  const roomName = roomNameStateEvent?.event?.content.name ?? 'NeoBoard';
-
-  const downloadUrl = useGenerateFile();
+  const { isDownloading, handleDownloadClick } = useWhiteboardDownload(onClick);
 
   return (
     <LoadingButton
-      component="a"
-      download={`${roomName}.nwb`}
-      loading={!downloadUrl}
-      href={downloadUrl}
-      onClick={onClick}
+      loading={isDownloading}
+      onClick={handleDownloadClick}
       startIcon={<FileDownloadIcon />}
-      target="_blank"
       variant="contained"
     >
       {children}
     </LoadingButton>
   );
 }
 
-function useGenerateFile() {
-  const [downloadUrl, setDownloadUrl] = useState<string>();
-  const whiteboardInstance = useActiveWhiteboardInstance();
-
-  useEffect(() => {
-    setDownloadUrl(undefined);
-
-    const whiteboardContent = whiteboardInstance.export();
-    const blob = new Blob([JSON.stringify(whiteboardContent)]);
-    const url = URL.createObjectURL(blob);
-
-    setDownloadUrl(url);
+function useWhiteboardDownload(onDownloadFinished: () => void) {
+  const { data: roomNameStateEvent } = useGetRoomNameQuery();
+  const roomName = roomNameStateEvent?.event?.content.name ?? 'NeoBoard';
+  const whiteboard = useActiveWhiteboardInstance();
+  const [isDownloading, setIsDownloading] = useState(false);
+  const widgetApi = useWidgetApi();
 
-    return () => {
-      URL.revokeObjectURL(url);
-    };
-  }, [whiteboardInstance]);
+  const handleDownloadClick = useCallback(async () => {
+    setIsDownloading(true);
+    const filename = `${roomName}.nwb`;
+    const whiteboardData = await whiteboard.export(
+      widgetApi.widgetParameters.baseUrl ?? '',
```

Comment:
`widgetApi.widgetParameters.baseUrl` have to be already defined there, better to throw an error as we do when image displays.

Reply:
If a download is not possible at all, we should eventually change the entire export modal to something like „Something is wrong, you cannot export your board“.

I will add it to the ticket for error handling. Would that be okay for you?

Reply:
It does now raise an error.